### PR TITLE
Add GetTicker to C2CX Client

### LINKED ISSUE #82: Add GetTicker to C2CX Client
https://api.c2cx.com/#ticker

### PR DESCRIPTION
**Pull request by Vavilen**
Fixes #82

### CODE REVIEW

File: exchange/c2cx/api.go [hunk 1]
```diff
@@ -481,6 +482,34 @@ func (c *Client) MarketSell(symbol TradePair, amount decimal.Decimal, customerID
 	return orderID, nil
 }
 
+type GetTickerResponse struct {
+	Code    int    `json:"code"`
+	Message string `json:"message"`
+	Data    struct {
+		Timestamp string  `json:"timestamp,omitempty"`
+		High      float64 `json:"high,omitempty"`
```
Comment (gz-c) [inline exchange/c2cx/api.go]: use `decimal.Decimal` instead of `float64`

File: exchange/c2cx/api.go [hunk 2]
```diff
@@ -481,6 +482,34 @@ func (c *Client) MarketSell(symbol TradePair, amount decimal.Decimal, customerID
...
+	params.Add("symbol", string(symbol))
+	data, err := c.get(getTicker, params)
+	if err != nil {
+		return nil, err
+	}
+	var getTickerResponse GetTickerResponse
+	if err = json.Unmarshal(data, &getTickerResponse); err != nil {
+		return nil, NewOtherError(err)
+	}
+	return &getTickerResponse, err
```
Comment (gz-c) [inline exchange/c2cx/api.go]: check getTickerResponse.Code for error before returning (see other methods for example)

File: exchange/c2cx/api.go [hunk 3]
```diff
@@ -481,6 +482,34 @@ func (c *Client) MarketSell(symbol TradePair, amount decimal.Decimal, customerID
 	return orderID, nil
 }
 
+type GetTickerResponse struct {
```
Comment (gz-c) [inline exchange/c2cx/api.go]: Make this `getTickerResponse` so it is not exported. We will return the `Data` struct instead

File: exchange/c2cx/api.go [hunk 4]
```diff
@@ -481,6 +482,34 @@ func (c *Client) MarketSell(symbol TradePair, amount decimal.Decimal, customerID
...
+	Code    int    `json:"code"`
+	Message string `json:"message"`
+	Data    struct {
```
Comment (gz-c) [inline exchange/c2cx/api.go]: Define this struct as a public struct in `c2cx.go`

File: exchange/c2cx/api.go (hunk 2 above)
Comment (gz-c) [inline exchange/c2cx/api.go]: return `&getTickerResponse.Data` instead

File: exchange/c2cx/api.go [hunk 5]
```diff
@@ -481,6 +482,34 @@ func (c *Client) MarketSell(symbol TradePair, amount decimal.Decimal, customerID
...
+		Timestamp string  `json:"timestamp,omitempty"`
```
Comment (gz-c) [inline exchange/c2cx/api.go]: Make this a `time.Time` and implement custom `UnmarshalJSON` and `MarshalJSON` to handle it.  See examples in `c2cx.go`.

**Review by gz-c — CHANGES_REQUESTED** [review]

File: exchange/c2cx/api.go [hunk 6]
```diff
@@ -481,6 +482,34 @@ func (c *Client) MarketSell(symbol TradePair, amount decimal.Decimal, customerID
 	return orderID, nil
 }
 
+type GetTickerResponse struct {
+	Code    int    `json:"code"`
+	Message string `json:"message"`
+	Data    struct {
+		Timestamp string           `json:"timestamp,omitempty"`
+		High      *decimal.Decimal `json:"high,omitempty"`
```
Comment (gz-c) [inline exchange/c2cx/api.go]: is the pointer and omitempty necessary?

Comment (Vavilen) [reply exchange/c2cx/api.go]: It allows to return empty `data` -  data:{} (not with zero values) as in the original c2cx response.

**Comment by gz-c** [comment]
needs updates

File: exchange/c2cx/c2cx.go [hunk 7]
```diff
@@ -431,3 +431,14 @@ type Orders struct {
 	Orders []Order `json:"orders"`
 	Page   int     `json:"page,omitempty"`
 }
+
+// TickerData includes pricing data for a give currency pair
+type TickerData struct {
+	Timestamp string           `json:"timestamp,omitempty"`
```
Comment (gz-c) [inline exchange/c2cx/c2cx.go]: is Timestamp measured in seconds or milliseconds?  Can you use a custom MarshalJSON and UnmarshalJSON to convert it to a time.Time? (see other examples in this file)

**Review by gz-c — CHANGES_REQUESTED** [review]

File: exchange/c2cx/c2cx.go [hunk 8]
```diff
@@ -434,11 +434,71 @@ type Orders struct {
...
+		t.Timestamp = time.Unix(ts, 0).UTC()
+	}
+	return nil
+}
+
+// MarshalJSON implements json.Marshaler
+func (t *TickerData) MarshalJSON() ([]byte, error) {
+	type rec struct {
+		Timestamp time.Time       `json:"timestamp"`
+		High      json.RawMessage `json:"high,omitempty"`
```
Comment (gz-c) [inline exchange/c2cx/c2cx.go]: why RawMessage? decimal was fine

**Review by gz-c — CHANGES_REQUESTED** [review]

File: exchange/c2cx/c2cx.go (hunk 8 above)
Comment (Vavilen) [reply exchange/c2cx/c2cx.go]: Original **data** JSON is:
```json
{  
    "timestamp":"1523336777",
    "high":43.7,
    "last":40.6,
    "low":39.6,
    "buy":40.6,
    "sell":41.0,
    "volume":5342.5395026767
}
```
But while decimal.Decimal field is marshalling, the String method is calling for the field type and the value becomes wrapped in the quotes.
So, this JSON now is looking like this:
```json
{
    "timestamp": "2018-04-10T05:14:44Z",
    "high": "43.7",
    "last": "40.6",
    "low": "39.6",
    "buy": "40.6",
    "sell": "41",
    "volume": "5350.9041135482"
}
```

If it's OK, I can return decimal.Decimal back.

Comment (gz-c) [reply exchange/c2cx/c2cx.go]: use custom MarshalJSON

File: exchange/c2cx/c2cx.go [hunk 9]
```diff
@@ -481,24 +481,10 @@ func (t *TickerData) UnmarshalJSON(b []byte) error {
...
-		High:      t.High,
-		Last:      t.Last,
-		Low:       t.Low,
-		Buy:       t.Buy,
-		Sell:      t.Sell,
-		Volume:    t.Volume,
-	}
-	return json.Marshal(tmp)
+	var actualValue = decimal.MarshalJSONWithoutQuotes
+	decimal.MarshalJSONWithoutQuotes = true
```
Comment (gz-c) [inline exchange/c2cx/c2cx.go]: dont change the global, this is a library

Comment (gz-c) [reply exchange/c2cx/c2cx.go]: the export value doesn't even matter, just leave it as decimal.Decimal

**Review by gz-c — CHANGES_REQUESTED** [review]

**Review by gz-c — APPROVED** [review]

**Pull request merged**